daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -818,31 +843,50 @@ private[lf] final class Compiler(
       tmpl: Template,
   ): (t.SDefinitionRef, SDefinition) =
     unlabelledTopLevelFunction2(t.ToContractInfoDefRef(tmplId)) { (tmplArgPos, mbKeyPos, env) =>
-      SBuildContractInfoStruct(
-        s.SEValue(STypeRep(TTyCon(tmplId))),
-        env.toSEVar(tmplArgPos),
-        t.AgreementTextDefRef(tmplId)(env.toSEVar(tmplArgPos)),
-        t.SignatoriesDefRef(tmplId)(env.toSEVar(tmplArgPos)),
-        t.ObserversDefRef(tmplId)(env.toSEVar(tmplArgPos)),
-        tmpl.key match {
-          case None =>
-            s.SEValue.None
-          case Some(tmplKey) =>
-            s.SECase(
-              env.toSEVar(mbKeyPos),
-              List(
-                s.SCaseAlt(
-                  t.SCPNone,
-                  let(env, translateExp(env.bindExprVar(tmpl.param, tmplArgPos), tmplKey.body)) {
-                    (keyPos, env) =>
-                      SBSome(translateKeyWithMaintainers(env, keyPos, tmplKey))
-                  },
-                ),
-                s.SCaseAlt(t.SCPDefault, env.toSEVar(mbKeyPos)),
-              ),
-            )
-        },
-      )
+      // We use a chain of let bindings to make the evaluation order of SBuildContractInfoStruct's arguments is
+      // independent from the evaluation strategy imposed by the ANF transformation.
+      let(env, s.SEValue(STypeRep(TTyCon(tmplId)))) { (typePos, env) =>
```

Comment:
:+1: 
Making the evaluation order of contract info components *not* be determined by function-arg evaluation order is a win.